### PR TITLE
Var swap

### PR DESCRIPTION
**Pull request by kevingivens**
Adding variance swap instrument and pricers.  Also BlackVolSurface/TermStructure which are needed for the pricing engines.  Two commits:  ff4ae77 Black Vol Surface, 71133f0 Variance Swap and Pricers

### CODE REVIEW

File: quantlib/math/_interpolations.pxd [hunk 1]
```diff
@@ -3,12 +3,38 @@ cdef extern from 'ql/math/interpolations/backwardflatinterpolation.hpp' namespac
...
         pass
 
+
 cdef extern from 'ql/math/interpolations/loginterpolation.hpp' namespace 'QuantLib':
 
     cdef cppclass LogLinear:
         pass
 
+
+cdef extern from 'ql/math/interpolations/linearinterpolation.hpp' namespace 'QuantLib':
```
[the remark this answers is not in the document]
Comment (kevingivens) [reply quantlib/math/_interpolations.pxd:13]: typo, fixed in 3576b7d

File: quantlib/termstructures/volatility/equityfx/black_variance_curve.pyx [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+                                               deref(reference_date._thisptr),
+                                               _dates,
+                                               black_vols,
+                                               deref(day_counter._thisptr),
+                                               force_monotone_variance,
+                                               ))
+    # TermStructure interface
+    @property
+    def day_counter(self):
+        """ ql lacks a copy constructor for DayCounter"""
```
[the remark this answers is not in the document]
Comment (kevingivens) [reply quantlib/termstructures/volatility/equityfx/black_variance_curve.pyx]: incorporated in 3576b7d

File: quantlib/termstructures/volatility/equityfx/black_variance_surface.pyx [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        for d in dates:
+            _dates.push_back(deref((<Date>d)._thisptr))
+
+        self._thisptr = shared_ptr[_bvts.BlackVolTermStructure](
+                                new _bvs.BlackVarianceSurface(
+                                             deref(reference_date._thisptr),
+                                             deref(cal._thisptr),
+                                             _dates,
+                                             strikes,
+                                             (<Matrix>black_vol_matrix)._thisptr,
```
[the remark this answers is not in the document]
Comment (kevingivens) [reply quantlib/termstructures/volatility/equityfx/black_variance_surface.pyx]: removed from 3576b7d

File: quantlib/termstructures/volatility/equityfx/black_variance_surface.pyx [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+                 vector[Real] strikes,
+                 Matrix black_vol_matrix,
+                 DayCounter dc,
+                 Extrapolation lower_extrap = Extrapolation.InterpolatorDefaultExtrapolation,
+                 Extrapolation upper_extrap = Extrapolation.InterpolatorDefaultExtrapolation,
+                 ):
+
+        cdef vector[_Date] _dates
+        for d in dates:
+            _dates.push_back(deref((<Date>d)._thisptr))
```
[the remark this answers is not in the document]
Comment (kevingivens) [reply quantlib/termstructures/volatility/equityfx/black_variance_surface.pyx]: added in 3576b7d

File: quantlib/termstructures/volatility/equityfx/_black_vol_term_structure.pxd [hunk 5]
```diff
@@ -9,28 +9,102 @@ from quantlib.termstructures._vol_term_structure cimport VolatilityTermStructure
...
 cdef extern from 'ql/termstructures/volatility/equityfx/blackvoltermstructure.hpp' namespace 'QuantLib':
 
-    cdef cppclass BlackVolTermStructure(VolatilityTermStructure):
-        BlackVolTermStructure(
-            Calendar& cal,
-            BusinessDayConvention bdc,
-            DayCounter& dc
-        )
-        Volatility blackVol(const Date& maturity,
+    cdef cppclass BlackVolTermStructure:
```
[the remark this answers is not in the document]
Comment (kevingivens) [reply quantlib/termstructures/volatility/equityfx/_black_vol_term_structure.pxd]: I added VolatilityTermStructure as a base class for BlackVolTermStructure in 3576b7d

**Review by dpinte — CHANGES_REQUESTED** [review]
@kevingivens There is one failure in the test suite. Can you look into it?

**Comment by thrasibule** [comment]
@kevingivens I've pushed the small fixes needed to get the test to run here: https://github.com/thrasibule/pyql/tree/varswap. Don't let your pr go to waste !

**Comment by kevingivens** [comment]
thanks @dpinte and @thrasibule.  I incorporated the changes.

File: quantlib/test/test_variance_swap.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+            if data['type'] == OptionType.Call:
+                call_strikes.append(data['strike'])
+                call_vols.append(data['v'])
+            elif data['type'] == OptionType.Put:
+                put_strikes.append(data['strike'])
+                put_vols.append(data['v'])
+            else:
+                raise ValueError("unknown option type")
+
+        #vols = Matrix(len(replicating_option_data)-1, 1)
```
Comment (dpinte) [inline quantlib/test/test_variance_swap.py:100]: can you remove commented code?

File: quantlib/test/test_variance_swap.py [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+            't' : 0.246575,
+            'v' : 0.20,
+            'result' : 0.04189,
+            'tol' : 1.0e-4,
+        }
+        self.spot.value = self.values['s']
+        self.q_rate.value = self.values['q']
+        self.r_rate.value = self.values['r']
+        self.ex_date = self.today + int(self.values['t']*365+0.5)
+        #self.factory = (MakeMCVarianceSwapEngine(stochProcess).
```
Comment (dpinte) [inline quantlib/test/test_variance_swap.py:50]: should this be part of another test? or just removed?

File: quantlib/termstructures/volatility/equityfx/black_variance_surface.pyx [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+                                             deref(cal._thisptr),
+                                             _dates,
+                                             strikes,
+                                             (black_vol_matrix)._thisptr,
+                                             deref(dc._thisptr),
+                                             lower_extrap,
+                                             upper_extrap,
+                                             ))
+
+    # TermStructure interface
```
Comment (dpinte) [inline quantlib/termstructures/volatility/equityfx/black_variance_surface.pyx:87]: Commented code should be removed or uncommented. Cany you check?

File: quantlib/termstructures/volatility/equityfx/black_variance_curve.pyx [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+        self._thisptr = shared_ptr[VolatilityTermStructure](
+                                   new _bvc.BlackVarianceCurve(
+                                               deref(reference_date._thisptr),
+                                               _dates,
+                                               black_vols,
+                                               deref(day_counter._thisptr),
+                                               force_monotone_variance,
+                                               ))
+    # TermStructure interface
```
Comment (dpinte) [inline quantlib/termstructures/volatility/equityfx/black_variance_curve.pyx:73]: Commented code should be removed or uncommented. Can you check?

**Review by dpinte — CHANGES_REQUESTED** [review]
@kevingivens look mostly good to me. Can you check the commented lines in the PR and decide if they need to be removed or uncommented?

File: quantlib/termstructures/volatility/equityfx/black_variance_curve.pyx (hunk 9 above)
Comment (kevingivens) [reply quantlib/termstructures/volatility/equityfx/black_variance_curve.pyx]: removed in 1244cdc

File: quantlib/termstructures/volatility/equityfx/black_variance_surface.pyx (hunk 8 above)
Comment (kevingivens) [reply quantlib/termstructures/volatility/equityfx/black_variance_surface.pyx]: removed in 1244cdc

File: quantlib/test/test_variance_swap.py (hunk 7 above)
Comment (kevingivens) [reply quantlib/test/test_variance_swap.py]: removed in 1244cdc.  I had trouble getting the MCVarianceSwapEngine factory to work.  I'm not sure factory methods like these are really needed in Python. I can add this later if we want to include it.

File: quantlib/test/test_variance_swap.py (hunk 6 above)
Comment (kevingivens) [reply quantlib/test/test_variance_swap.py]: removed in 1244cdc

**Review by dpinte — APPROVED** [review]
LGTM!

**Pull request merged**